crossplane/crossplane
# Generate TLS certificates for ESS and use them

### Description of your changes

In this change we started to generate TLS certificates and mount them to XP and provider pods when **the External Secret Stores alpha feature is enabled**. These certificates will be used for pluggable secret store authentications. Design details can be found in this [doc](https://github.com/crossplane/crossplane/blob/master/design/one-pager-pluggable-secret-stores.md). 



Fixes https://github.com/upbound/team-control-planes/issues/966

I have:


### How has this code been tested
E2E test with following steps:

1) Create a local kind cluster
```
/cluster/local/kind.sh up
```
2) Prepare Vault as described [here](https://github.com/crossplane/docs/blob/master/content/knowledge-base/integrations/vault-as-secret-store.md#prepare-vault)
3) Build and install Crossplane on this branch
```
make
HELM3_FLAGS="--set args={--debug,--enable-external-secret-stores} " ./cluster/local/kind.sh helm-install crossplane-system
```
4) Install a custom provider-gcp and create provider config
```
kubectl apply -f provider-gcp-ess.yaml
```
5) Install storeconfigs
```
kubectl apply -f storeconfig-xp.yaml -f storeconfig-gcp.yaml
```
6) Install [ESS plugin Vault](https://github.com/crossplane-contrib/ess-plugin-vault)
```
helm upgrade --install ess-plugin-vault oci://xpkg.upbound.io/crossplane-contrib/ess-plugin-vault --namespace crossplane-system -f values.yaml
```
7) Install vaultconfig
```
kubectl apply -f vaultconfig.yaml
```
8) Install XRD and composition:
```
kubectl apply -f xrd.yaml -f composition.yaml
```
9) Install claim
```
kubectl apply -f claim.yaml
```
10) Portforward Vault service and verify that secrets are written
```
kubectl port-forward svc/vault 8200:8200 -n vault-system
```
11) Delete claim and verify that secrets are deleted from Vault
```
kubectl delete -f claim.yaml
```

All the files mentioned above can be found [here](https://gist.github.com/ezgidemirel/190721cbc6e3b3449697888dbc001911).

Also tested two upgrade scenarios where the feature is enabled/disabled in the previous XP version and verified that k8s secrets with certificates are created, and mounted to XP and provider pods after the upgrade as expected. 





[contribution process]: https://git.io/fj2m9
